### ISSUE TITLE
Bug: no-useless-assignment false positive in vue sfc

### ISSUE THREAD

**Issue by Trinovantes** [eslint/eslint#20501]
### Environment

Node version: v22.19.0
npm version: v11.7.0
Local ESLint version: v10.0.0 (Currently used)
Global ESLint version: Not found
Operating System: linux 6.6.87.2-microsoft-standard-WSL2

### What parser are you using?

@typescript-eslint/parser

### What did you do?

https://github.com/Trinovantes/eslintjs10vuebug

Running `pnpm eslint` will trigger a `no-useless-assignment` error

Running `pnpm add -D @eslint/js@^9` first will work as expected

### What did you expect to happen?

No `no-useless-assignment` error

### What actually happened?

```
/home/stephen/eslintjs10vuebug/src/MyComponent.vue
  4:7  error  This assigned value is not used in subsequent statements  no-useless-assignment
```

### Link to Minimal Reproducible Example

https://github.com/Trinovantes/eslintjs10vuebug

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
@Trinovantes thanks for the repro!

Does the same problem occur with ESLint v9 when you enable the `no-useless-assignment` rule?

(note that in ESLint v9 this rule wasn't included in the `recommended` config)

**Comment by mdjermanovic** [comment]
I checked what's happening when the rule evaluates this code, and the problem seems to be caused by results of scope analysis.

This is the code from the repro:

```vue
<script lang="ts" setup>
import { ref } from 'vue'

const shouldShow = ref(false)
</script>

<template>
    <div v-if="shouldShow">
        Hello World
    </div>
</template>
```

Variable `shouldShow` has two references:

1. The first one is a write reference that corresponds to the variable initialization, and it looks as expected.
2. The second one is a read reference, and it has `vueUsedInTemplate: true` property, so I believe it corresponds to the reference in the template. But, its `identifier` is the `shouldShow` identifier in `const shouldShow = ref(false)`, and therefore the rule doesn't consider it as been evaluated after the initialization and thus reports the initial assignment as unused.

@ota-meshi is this the expected result of scope analysis?

**Comment by ota-meshi** [comment]
I think the cause of the false positive is that vue-eslint-parser doesn't mark `eslintUsed` for now.

I previously considered marking variables with `eslintUsed`, but I found that it was not useful for the popular rules in tseslint at the time. At the time, I made them compatible in a different way instead of `eslintUsed`, but I think that marking rules with `eslintUsed` is necessary for compatibility with the `no-useless-assignment` rule.

https://github.com/typescript-eslint/typescript-eslint/issues/4508#issuecomment-1030513892

If you want to track the issue, please file an issue with vue-eslint-parser.

https://github.com/vuejs/vue-eslint-parser

**Comment by mdjermanovic** [comment]
@ota-meshi thanks for the info!

@Trinovantes can you please file an issue with vue-eslint-parser?

We'll leave this issue open for visibility until the problem is resolved.

**Comment by Trinovantes** [comment]
This issue is now resolved with vue-eslint-parser 10.4.0
